docs/tools/firmware.sh

```diff
@@ -621,35 +918,30 @@ case "$os" in
 					unmount_macos_and_cleanup
 					exit 1
 				fi
-				python3 "$0" ${fwlocation} ${workdir}/firmware-renamed.tar ${verbose} || (echo -e "\nCouldn't extract firmware. Try running the script again. If error still persists, try restarting your Mac and then run the script again, or choose some other method." && unmount_macos_and_cleanup && exit 1)
-				sudo tar ${verbose} -xC /lib/firmware/brcm -f ${workdir}/firmware-renamed.tar
-				echo "Reloading Wi-Fi and Bluetooth drivers"
-				sudo modprobe -r ${verbose} brcmfmac_wcc || true
-				sudo modprobe -r ${verbose} brcmfmac || true
-				sudo modprobe ${verbose} brcmfmac || true
-				sudo modprobe -r ${verbose} hci_bcm4377 || true
-				sudo modprobe ${verbose} hci_bcm4377 || true
+				create_firmware_archive "${fwlocation}" "${workdir}/firmware-renamed.tar" || (echo -e "\nCouldn't extract firmware. Try running the script again. If error still persists, try restarting your Mac and then run the script again, or choose some other method." && unmount_macos_and_cleanup && exit 1)
+				sudo tar ${verbose} -xC /lib/firmware/brcm -f "${workdir}/firmware-renamed.tar"
+				reload_kernel_modules
 				echo "Cleaning up"
 				unmount_macos_and_cleanup
 				echo "Done!"
 				;;
 			(3)
 				# Detect whether curl and dmg2img are installed
 				echo -e "\nChecking for missing dependencies"
-				curl_check
-				dmg2img_check
+				curl --version >/dev/null 2>&1 || install_package curl
+				dmg2img >/dev/null 2>&1 || install_package dmg2img
 				cleanup_dmg () {
-					sudo rm -r ${verbose} ${workdir}
-					sudo umount ${verbose} ${loopdevice}
-					sudo rm -r ${verbose} ${imgdir}
-					sudo losetup -d /dev/${loopdev}
+					sudo rm -r ${verbose} "${workdir}"
+					sudo umount ${verbose} "${loopdev_partition}"
+					sudo rm -r ${verbose} "${imgdir}"
+					sudo losetup -d /dev/"${loopdev}"
 				}
 
 				echo -e "\nDownloading macOS Recovery Image"
-				workdir=$(mktemp -d)
+				workdir=$(mktemp -d "$HOME/.tmp.XXXXXX")
```

Comment:
Don’t make it hidden. In case the script halts in between, such as user presses control+c, Mac discharges etc, the folder is visible to the user and he can remove it manually.